### ISSUE TITLE
Changes to handlers for node is not being picked up by live reload

### ISSUE THREAD

**Issue by vietanhtran16** [kettanaito/msw#1389]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Node.js version

v14.18.2

### Reproduction repository

https://github.com/vercel/next.js/tree/canary/examples/with-msw

### Reproduction steps

- Run `yarn dev`
- Makes changes to handler for `https://my.backend/book`
- Refresh the page
- Changes made to handler is not being reflected

### Current behavior

In `msw@0.41.1`, this version, changes made to handler was being reflected with live reload. After this version, this does not work anymore. Looks like a regression in versions after this

### Expected behavior

Changes made to handler should be reflected without having to stop and start the server like in `0.41.1` version

**Comment by kettanaito** [comment]
Hey, @vietanhtran16.

I believe you're referring to the handler for [this server-side call](https://github.com/vercel/next.js/blob/a4d907a564e0f73adcaeaa684a00a3faefc5a678/examples/with-msw/pages/index.tsx#L40). Since that call happens in the server process of Next, the changes you make to your handlers won't propagate there. This is one of Next's limitations and there's nothing MSW can do about it. This is how it always worked. You must be referring to something else saying this is a regression. You have to restart the Next server if you make a change to your handlers that affect server-side network calls. 

For the `/review` endpoint, on the other hand, you will see the changes apply immediately with live reloading because you're making [that request on the client-side](https://github.com/vercel/next.js/blob/a4d907a564e0f73adcaeaa684a00a3faefc5a678/examples/with-msw/pages/index.tsx#L13).

**Comment by vietanhtran16** [comment]
Hey @kettanaito, I have created [a repo](https://github.com/vietanhtran16/next.js/tree/msw-0.41.1) that reproduce that live reload worked with msw@0.41.1 for handlers on server side process of NextJs. That is why I think it is a regression since it used to work. Could we reopen this issue please? 

Steps to reproduce:
- Checkout the above branch on my fork
- `cd examples/with-msw/`
- `yarn install`
- `yarn dev`
- Trying changing mock response for `https://my.backend/book` (called in `getServerSideProps`) and `https://my.backend/user` (called when clicking on `Load Reviews` => request to API Routes => which calls `https://my.backend/user`)

- When you refresh the page, you should be able to see the updated mock response.

I have also attached a GIF as example
![Kapture 2022-09-21 at 11 48 13](https://user-images.githubusercontent.com/23372284/191384772-ec235b27-aa0a-4319-91a7-d2c476458aa7.gif)

**Comment by kettanaito** [comment]
I'm suspecting changes on the Next's side. MSW hasn't introduced any changes that would break this, afaik. Thank you for creating a reproduction repository! I may look into it once I have a spare minute. Meanwhile, contributors are welcome!

**Comment by chrisb2244** [comment]
@vietanhtran16 I'm curious about this, which version of msw did you use for the 'not working' version? (I see 0.41.1 for working).
Do you also know if you're using different versions of Next?

I had to change the way I was importing MSW for 0.47(? I think), and briefly pinned 0.46.something(?), but went back to `@latest` once I worked out how to correctly import... At that point, I hadn't gotten anything I cared about working, so I don't know if I had reloadable handlers or not, but changes to import method strike me as a possible cause?

I'll note that in MSW 0.47.3 and 0.47.4, hot reloading reimports the handlers (e.g. the console log here is triggered in _app.tsx)
```
if (process.env.PLAYWRIGHT === '1') {
  console.log('loading mocks')
  import('../../playwright/mocks').then(({initMocks}) => {
    initMocks().then(mock => mock.resetHandlers())
  })
}
```
but that no combinations of initMocks() and resetHandlers() allowed me to get the handlers to update (I can also console log the serverStart call, etc, and see it being called again, but it might not be the same call that is in fact handling things? Is it possible I have a number of irrelevant MSW processes and the one that is actually handling requests is just the first one loaded with the node process?

**Comment by chrisb2244** [comment]
[v0.47.0](https://github.com/mswjs/msw/releases/tag/v0.47.0) was probably the latest as of 5th September when you reported this - it was released the previous day - did you just update?

**Comment by kettanaito** [comment]
>  Is it possible I have a number of irrelevant MSW processes

This sounds very strange. MSW does not spawn processes. If you have multiple seemingly MSW-related processes, those are either your app or your tests spawning themselves more than necessary.

**Comment by chrisb2244** [comment]
Sorry, I wasn't saying I _did_ have multiple processes - only wondering if the `serverSetup(...handlers)` calls I'm seeing are not related to the msw server that's actually handling requests. I guess I can go edit `handlers.ts` a bunch of times, see the application reload, and see if anything changes via `top` or something...

If you think that's unlikely/impossible, then I guess the problem is somewhere else.

**Comment by vietanhtran16** [comment]
Hey @chrisb2244, any version after `0.41.1` of `msw` didnt work for me back then. I think the newest version that I tested against was `0.46.0`, I can test again to see if it is still an issue.

In regards to `nextjs` version, in the [reproducable repo](https://github.com/vietanhtran16/next.js/tree/msw-0.41.1), I am using `latest` nextjs. In my work's repo, I am on `12.2.0`


From your previous answer, is this how you got hot reload mock to work?
```
import('../../playwright/mocks').then(({initMocks}) => {
    initMocks().then(mock => mock.resetHandlers())
  })
```

**Comment by chrisb2244** [comment]
@vietanhtran16 Not in a useful way - if you do that then you can see the functions being executed again by the hot-reload (e.g. if you put some `console.log('Loading handlers')` in `initMocks` or similar), but it doesn't actually change the behaviour of the handlers (easily exhibited by putting a `console.log('this is a test')` in a handler, then changing the text logged - the hot-reload for Next occurs, the "Loading handlers" message appears, but triggering the MSW handler will still log the previous text (no idea why).

My previous hypothesis was some sort of unused worker vs active worker, but seems like kettanaito said that wasn't possible, and I was unable in the time since then to get any `top` running in the same instance as my processes (Docker... :/ ).
I might try with a local copy (not Docker container) and see if it's more readily debuggable, but I don't currently have a new hypothesis as to what's happening.

**Comment by vietanhtran16** [comment]
> @vietanhtran16 Not in a useful way - if you do that then you can see the functions being executed again by the hot-reload (e.g. if you put some `console.log('Loading handlers')` in `initMocks` or similar), but it doesn't actually change the behaviour of the handlers (easily exhibited by putting a `console.log('this is a test')` in a handler, then changing the text logged - the hot-reload for Next occurs, the "Loading handlers" message appears, but triggering the MSW handler will still log the previous text (no idea why).
> 
> My previous hypothesis was some sort of unused worker vs active worker, but seems like kettanaito said that wasn't possible, and I was unable in the time since then to get any `top` running in the same instance as my processes (Docker... :/ ). I might try with a local copy (not Docker container) and see if it's more readily debuggable, but I don't currently have a new hypothesis as to what's happening.

I see yeah I have tried to use console.log to debug but couldn't figure out why it is not picking up the new handlers.

@kettanaito I just tested with the latest version `1.0.0` but it is still not working for server side handlers on nextjs (working fine for browser side handlers). We are stuck at version `0.41.1` at the moment due to this and would like to be able to keep upgrading to newer version. Would you be able to spend some time to look at this or point me where in the code I should look at so I can try to figure it out? My suspicion is that one of the PRs after `0.41.1` broke live reloading for server side handlers but couldn't pinpoint what is causing it. Happy to collaborate on this. Thank you 👍

**Comment by vietanhtran16** [comment]
Hey @kettanaito would love to get your thought on this issue. Thank you 👍

**Comment by vietanhtran16** [comment]
I did some more digging and actually live reload still works with msw@0.43.1 (not 0.41.1 as I mentioned before) for handlers on server side process of NextJs.

As soon as I tried to upgrade to `0.44.0`, it stopped working. I look at the changes between the [2 versions](https://github.com/mswjs/msw/compare/v0.43.1...v0.44.0) and nothing really pops out as a potential cause of this issues. I will spend a bit more time digging but if you noticed something that can cause problem between those 2 versions, please let me know @kettanaito. Thank you

**Comment by kettanaito** [comment]
Hey, @vietanhtran16. Thanks for all the insight into the issue. 

I've noticed that Next has changed significantly over the past few months, even the older versions of Next (12 is it?). For example, when trying to run the existing `with-msw` template designed for Next 12, I couldn't get it to work. The version of MSW is pinned there, so I assume something has changed in how Next works. I have a suspicion to believe they've shipped some internal improvements in v12 to incrementally adopt them in v13 as well. I would love for the Next team member to prove me wrong. 

Frankly, I see little I can suggest. HMR can really stop working only for one reason: the changed module is not in the update path. From what I can see, the handlers module should be in the update path as the import map is clear:

```
_app.js
  - import mocks/server.js
    - import handlers.js
```

Now, why it's different on different versions of MSW that is a pickle. We can affect this with the library's internal structure to some degree but I don't think we've shipped any changes between 0.43 and 0.44 that would cause this. I'm happy to be proven wrong on this. 

It's also important to define what "stopped working" means. If you modify your handlers and the changes are not reflected on the server-side behavior of Next, the issue may not even be in HMR but in how Next handles the re-evaluation of modules upon changes and whether the server process is even restarted. 

I don't have the time capacity to debug this at the moment. For anybody willing to dive deeper, I would recommend the following:

- Try to confirm/deny that the hot update path reaches the `handlers.js` module. It'd be quite a different issue if it does/doesn't.
- Check the `process.pid` on the server-side and whether it changes between hot updates.
- Find the farthest point in the update path that still gets updated. Maybe we can deduce from there. 

Also, be mindful of the importance of this issue since it's pretty much obsolete in Next 13. The architecture there is completely different, I wrote a little about it on Twitter. We will have to (hopefully) collaborate with the Next team because at the moment of writing this Next provides no means to execute server-side logic that'd affect all server-side components.

**Comment by vietanhtran16** [comment]
Thanks for sharing your thoughts on this @kettanaito.

My colleague and I did some debugging on this yesterday and we found something interesting.

Dependencies' versions that we had
- `msw@1.2.1`
- `@mswjs/interceptors@0.22.12`

We had the normal set up with this `mocks/index.ts` file 
![image](https://user-images.githubusercontent.com/23372284/235797534-e1861fcc-ef3e-409e-8258-6846ee4ed8f7.png)
Which is imported into `_app.tsx`
![image](https://user-images.githubusercontent.com/23372284/235797582-89b714bb-d83f-46cd-a78c-d63e2420bbba.png)

We added some console log into msw `init` method of `SetUpServerApi` in [here](https://github.com/mswjs/msw/blob/main/src/node/SetupServerApi.ts#L63). Interestingly, `this.currentHandlers` outside of the `this.interceptor.on("request")` call back always has the updated handlers as we change them while `this.currentHandlers` inside the interceptor always has old handlers. This feels like the very first `this.interceptor.on("request")` is always used and not cleaned up properly. Based on this, is there anything you can think of that can cause this issue? Thank you so much
![image](https://user-images.githubusercontent.com/23372284/235797789-6df05602-0b19-4e50-ae96-d139ae8a7272.png)

**Comment by mizdra** [comment]
Here is a workaround I found.

```ts
import {server} from './server'

export function initMSW() {
  if (typeof window === 'undefined') {
    if ('listeningServer' in globalThis) {
      ;(globalThis as any).listeningServer.close()
    }
    server.listen({
      onUnhandledRequest: 'bypass',
    })
    ;(globalThis as any).listeningServer = server
    console.log('[MSW] Mocking enabled.')
  } else {
    import('./worker').then(({worker}) => {
      return worker.start({
        onUnhandledRequest: 'bypass',
      })
    })
  }
}
```

## Description
Next.js live reload simply re-evaluates the files that have been modified from the top. However, the evaluation is performed on the execution environment in which the previous version of the code was evaluated. In other words, when you do a live reload, any resources that were not released in the previous version are carried over to the next version.

In other words, if a server is listening in the previous version and the application is live-reloaded without closing it, the previous server is still alive and its handler is used in the next version. This is the cause of the problem.

To solve this problem, simply close the previous version of the server when the application is live-reloaded.

**Comment by easynowbaby** [comment]
@mizdra Thanks a lot! It doesn't reload handlers when live-reloading but it works when the page is refreshed which is still massively helpful

**Comment by kettanaito** [comment]
There's a lot of research and changes that have happened since this issue has been reported. Please follow https://github.com/mswjs/examples/pull/101, which is the official Next.js + MSW example where the Next.js team and I are tackling the remaining DX issues. 

I'm closing this.
